### PR TITLE
Decode and verify

### PR DESCRIPTION
**Pull request by rohe**
Thanks for contributing to this library! Please include the following check
list in your pull request submission (you can delete this message). If your
changes don't need a change log or documentation update, please ignore this.

- [x] Any changes relevant to users are recorded in the `CHANGELOG.md`.
- [ ] The documentation has been updated, if necessary.

---

Working with the OP test tools a misbehaviour in pyoidc became obvious.
In protocol messages the value None on a parameter means that that parameter should just be ignored.
This is not true for Claims and IDToken's. 
None values on parameters in either of these messages should be regarded as errors.

There also where some transformation errors when creating Message instances.
Most notably a non-integer was accepted as a value to a parameter marked as having integer value.

And lastly a problem with Python was discovered.
For some reason isinstance(True, int) == True
So, I had to add some code to work around this.

### CODE REVIEW

File: src/oic/oic/message.py [hunk 1]
```diff
@@ -546,6 +546,10 @@ def verify(self, **kwargs):
                     except ValueError:
                         raise VerificationError("Birthdate format error", self)
 
+        for val in self.values():
```
Comment (lwm) [inline src/oic/oic/message.py]: Could be `if any(val is None for val in self.values()):`

File: src/oic/utils/keyio.py [hunk 2]
```diff
@@ -333,6 +333,16 @@ def kids(self):
         self._uptodate()
         return [key.kid for key in self._keys if key.kid != ""]
 
+    def remove_outdated(self, after):
```
Comment (lwm) [inline src/oic/utils/keyio.py]: Feel like adding a docstring?

File: src/oic/utils/keyio.py [hunk 3]
```diff
@@ -333,6 +333,16 @@ def kids(self):
...
+        now = time.time()
+        _kl = []
+        for k in self._keys:
```
Comment (lwm) [inline src/oic/utils/keyio.py]: With my 'pythonic' hat on:

``` python
try:
    if not k.inactive_since + after < now:
        _kl.append(k)
except WhateverErrorThatIs:
    pass
```

File: src/oic/utils/keyio.py [hunk 4]
```diff
@@ -396,6 +407,7 @@ def __init__(self, ca_certs=None, verify_ssl=True, keybundle_cls=KeyBundle):
         self.ca_certs = ca_certs
         self.verify_ssl = verify_ssl
         self.keybundle_cls = keybundle_cls
+        self.remove_after = remove_stale_after
```
Comment (lwm) [inline src/oic/utils/keyio.py]: The naming is different, by choice or? The rest are the same.

File: src/oic/utils/keyio.py [hunk 5]
```diff
@@ -777,6 +789,39 @@ def get_issuer_keys(self, issuer):
...
+
+        # The set of issuers MUST be the same
+        if set(self.keys()) != set(other.keys()):
+            return False
+
+        # Keys per issuer must be the same
+        for iss in self.keys():
+            sk = self.get_issuer_keys(iss)
+            ok = other.get_issuer_keys(iss)
+            if len(sk) != len(ok):
```
Comment (lwm) [inline src/oic/utils/keyio.py]: The comment above suggests the key *values* must be the same, but only length or?

File: src/oic/utils/keyio.py [hunk 6]
```diff
@@ -777,6 +789,39 @@ def get_issuer_keys(self, issuer):
...
+                return False
+
+            for k in sk:
```
Comment (lwm) [inline src/oic/utils/keyio.py]: Could be: `if not any(k in ok for k in sk):`

File: src/oic/utils/keyio.py [hunk 7]
```diff
@@ -777,6 +789,39 @@ def get_issuer_keys(self, issuer):
...
+                if k not in ok:
+                    return False
+
+        return True
+
+    def remove_outdated(self):
```
Comment (lwm) [inline src/oic/utils/keyio.py]: Oh, another method with the same name? Now I demand a docstring :)

File: src/oic/utils/keyio.py [hunk 8]
```diff
@@ -777,6 +789,39 @@ def get_issuer_keys(self, issuer):
...
+        for iss in list(self.keys()):
```
Comment (lwm) [inline src/oic/utils/keyio.py]: Is the `list()` wrapper necessary?

File: src/oic/oauth2/message.py [hunk 9]
```diff
@@ -431,6 +431,12 @@ def _add_value(self, skey, vtyp, key, val, _deser, null_allowed):
         else:
             if val is None:
                 self._dict[skey] = None
+            elif isinstance(val, bool):
+                if vtyp is bool:
```
Comment (lwm) [inline src/oic/oauth2/message.py]: Would prefer to see `if isinstance(vtyp, bool)`.

File: src/oic/oauth2/message.py [hunk 10]
```diff
@@ -439,6 +445,15 @@ def _add_value(self, skey, vtyp, key, val, _deser, null_allowed):
                         val = _deser(val, sformat="dict")
                     except Exception as exc:
                         raise DecodeError(ERRTXT % (key, exc))
+                elif vtyp == int:
```
Comment (lwm) [inline src/oic/oauth2/message.py]: Again, would prefer an `isinstance` call.

File: src/oic/oauth2/message.py [hunk 11]
```diff
@@ -439,6 +445,15 @@ def _add_value(self, skey, vtyp, key, val, _deser, null_allowed):
...
+                    try:
+                        self._dict[skey] = int(val)
+                    except (ValueError, TypeError):
```
Comment (lwm) [inline src/oic/oauth2/message.py]: I don't see why we should ignore `TypeError` here? If your only wish is to pass a message with more information you can do `except (ValueError, TypeError) as exc:  raise exc(msg)`.

File: src/oic/oic/message.py (hunk 1 above)
Comment (rohe) [inline src/oic/oic/message.py]: OK

File: src/oic/utils/keyio.py (hunk 2 above)
Comment (rohe) [inline src/oic/utils/keyio.py]: Always ...

File: src/oic/utils/keyio.py (hunk 3 above)
Comment (rohe) [inline src/oic/utils/keyio.py]: Can buy that

File: src/oic/utils/keyio.py (hunk 4 above)
Comment (rohe) [inline src/oic/utils/keyio.py]: By choice but I guess mostly of historic reasons. 
So should possibly make them the same.

File: src/oic/utils/keyio.py (hunk 5 above)
Comment (rohe) [inline src/oic/utils/keyio.py]: The set of keys must be that same which means that the number of keys must be the same.
That's a MUST but not sufficient as the lines below show.
It's an optimisation. If the number of keys are different then I don't have to check the
individual keys.

File: src/oic/utils/keyio.py (hunk 6 above)
Comment (rohe) [inline src/oic/utils/keyio.py]: Could ..

File: src/oic/utils/keyio.py (hunk 7 above)
Comment (rohe) [inline src/oic/utils/keyio.py]: Sure, different classes so slightly different even though the basic task is the same.

File: src/oic/utils/keyio.py (hunk 8 above)
Comment (rohe) [inline src/oic/utils/keyio.py]: Probably not. But I've been bitten by not using it so many times now that it's become a default.

File: src/oic/oauth2/message.py (hunk 9 above)
Comment (rohe) [inline src/oic/oauth2/message.py]: Hmm

File: src/oic/oauth2/message.py (hunk 11 above)
Comment (rohe) [inline src/oic/oauth2/message.py]: I'm not ignoring TypeError. I'm just raising a ValueError if for some reason 'val' can not be 
cast to an int.

**Comment by rohe** [comment]
Thanks for the serious review Luke! Much appreciated !

File: src/oic/oauth2/message.py (hunk 9 above)
Comment (rohe) [inline src/oic/oauth2/message.py]: vtyp is not an instance of bool. It is the type bool so doing an isinstance test won't work.

File: src/oic/oauth2/message.py (hunk 10 above)
Comment (rohe) [inline src/oic/oauth2/message.py]: Same as above, vtyp is a type not an instance of a type.

**Comment by rohe** [comment]
@lwm if you're OK with the changes I made please merge till PR.
Time to release a new (minor) version ?

**Comment by lwm** [comment]
> Time to release a new (minor) version ?

If we make a next release, [it will need to be major](https://github.com/OpenIDC/pyoidc/blob/master/CHANGELOG.md#01100-unreleased). Yes, it's probably a good idea to do this now.

File: tests/test_keyio.py [hunk 12]
```diff
@@ -430,3 +431,44 @@ def test_get_signing_key_use_undefined():
...
+    assert len(_interm) == 4
+
+    # Now mark the keys to be inactivated
+    _now = time.time()
+    for k in keyjar.get_issuer_keys(''):
+        if k.kid in _old:
+            if not k.inactive_since:
+                k.inactive_since = _now
+
+    time.sleep(2)
```
Comment (lwm) [inline tests/test_keyio.py]: Oh! This is generally a bad idea. The slower the tests get, the more likely we are not to run them. I would absolutely recommend using [freezegun](https://github.com/spulec/freezegun) and a [manual tick](https://github.com/spulec/freezegun#manual-ticks) (which bumps time along instantly!).

**Comment by schlenk** [comment]
We could always still do a minor release. Its just the number in the changelog... (and next time we should probably not set any version number in the changelog for unreleased until we decide on the final number and release it).

**Comment by lwm** [comment]
> We could always still do a minor release.
> next time we should probably not set any version number in the changelog for unreleased

I think we have breaking changes included in the latest pull requests though? I think we named the new version number because there was an incompatible change, so it had to be this major bump. Could be wrong ...

**Comment by schlenk** [comment]
yes, probably right.

File: tests/test_keyio.py (hunk 12 above)
Comment (lwm) [inline tests/test_keyio.py]: @rohe, I'll add this change the to PR now and merge soon after.

**Comment by lwm** [comment]
:beers: :beers: :beers:

**Pull request merged**